From 8f2f91cb507dabd89f87160e4922f673cf085dbb Mon Sep 17 00:00:00 2001
From: Andrew Kelley <andrew@ziglang.org>
Date: Mon, 8 Sep 2025 18:18:27 -0700
Subject: [PATCH 100/133] std.Io.Writer.writeSliceEndian: add compile error
X-Developer-Signature: v=1; a=openpgp-sha256; l=884; i=xtex@astrafall.org;
 h=from:subject; bh=Me51jatG4KVVrwMLWHb+XSpyMo+6aZnrj5zrLAjJnvo=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78rhWij7/vvzFw7PLGg+w683cwLN1vlN6/7In0u7nL
 0wWMfz2p2MiC4MYF4OlmCJLkWGDN6tOOr/osnJZmDmsTCBDpEUaGBgYGFgY+HIT80qNdIz0TLUN
 9QwNdYx0jBi4OAVgqsv/MPxTO3SqTmT1jO0+rCtE/ga6hJW2LuiuKul7Y1aQLz7r0a6zjAxPH5r
 Jv5+yO7ErZaX8EbdH4mpGqgqlZyb6fo9edSjcWYMfAA==
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

check when an auto-layout struct is attempted to be memory reinterpreted
and written out. it would be writing undefined memory

(cherry picked from commit eac2bbfec9f99c27a886f4842cb9ca42584607f3)
---
 lib/std/Io/Writer.zig | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/lib/std/Io/Writer.zig b/lib/std/Io/Writer.zig
index 2bb86fd219fe..53c70b2209af 100644
--- a/lib/std/Io/Writer.zig
+++ b/lib/std/Io/Writer.zig
@@ -864,6 +864,11 @@ pub inline fn writeSliceEndian(
     slice: []const Elem,
     endian: std.builtin.Endian,
 ) Error!void {
+    switch (@typeInfo(Elem)) {
+        .@"struct" => |info| comptime assert(info.layout != .auto),
+        .int, .@"enum" => {},
+        else => @compileError("ill-defined memory layout"),
+    }
     if (native_endian == endian) {
         return writeAll(w, @ptrCast(slice));
     } else {
-- 
2.52.0

